[aotd] capture rrelu_with_noise noise mutation

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138503

Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

IvanKobzarev committed:
[aotd] capture rrelu_with_noise noise mutation

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/functorch/BatchRulesBinaryOps.cpp:
```diff
@@ -285,6 +285,44 @@ static void fill__Tensor_batch_rule(
+  auto [noise_value, noise_bdim] = unwrapTensorAtLevel(noise, cur_level);
+  TORCH_CHECK(!noise_bdim.has_value(), "vmap: Attempted to vmap over 'noise' in torch.rrelu_with_noise. This is not supported.");
+  auto res = rrelu_with_noise_batch_rule(self_value, self_bdim, noise_value, noise_bdim, lower, upper, training, std::move(generator));
+  return makeBatched(std::get<0>(res), std::get<1>(res), cur_level);
```
cc @zou3519 does this updated vmap rule look ok?

zou3519 replied:
Why did we need to change the vmap rule?

zou3519 replied:
Okay I see the mutation

bdhirsh reviewed (approved):
lgtm, although:

(1) I'll wait for Richard to comment on the batching rule

(2) there's an XLA failure, so you're probably going to need an XLA-side patch. @qihqi @miladm who from the XLA side can help coordinate?

From the error, it looks like you need to remove `const` from the `noise` argument in the XLA lowering here: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_xla_type.cpp#L3022

IvanKobzarev commented:
xla PR https://github.com/pytorch/xla/pull/8309 
b2430c10302a82bc56060f0f6944cf28afabd1a7

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/BatchRulesBinaryOps.cpp:
```diff
@@ -285,6 +285,44 @@ static void fill__Tensor_batch_rule(
+  return std::make_tuple(ret, 0, noise_, 0);
+}
+
+static Tensor rrelu_with_noise_batch(
```
Is this operation deterministic?

IvanKobzarev replied:
No, the noise is random uniform  on every call.

zou3519 reviewed (approved):
batching rule looks fine to me

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] capture rrelu_with_noise noise mutation"


Issues:
https://github.com/pytorch/pytorch/issues/136784
https://github.com/pytorch/pytorch/issues/135083


rrelu_with_noise mutates noise.
This was not reflected in schema.

1/ Changing schema that noise is mutated

2/ Turning on generation of functional variant of the op

3/ Adding derivatives rule for rrelu_with_noise_functional, as it has different schema

4/ Removing rrelu_with_noise from BatchRulesBinaryOps:
- because rrelu_with_noise is not compatible as a mutation op.
- rrelu_with_noise_functional returns Tuple[Tensor, Tensor], while BatchRuleBinaryOps macroses support only Tensor return type.

Testing:
```
python test/functorch/test_aotdispatch.py -k rrelu_noise_mutation
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]
